File: dangerstep.py
import numpy as np

# Define the game size
WIDTH = 80
HEIGHT = 31

class DangerStepGame:
    def __init__(self):
        self.grid = np.random.randint(1, 10, size=(HEIGHT,WIDTH))
        self.player_position = (HEIGHT // 2, WIDTH // 2)
        self.grid[self.player_position] = 0  # Start position is marked with 0
        self.score = 0
        self.path = set([self.player_position])
        self.game_over = False

    def move(self, direction):  
        if self.game_over:
            return self.score, True

        # Define the movement vectors
        movements = {
            'u': (-1, 0), 'd': (1, 0), 'l': (0, -1), 'r': (0, 1),
            'n': (-1, -1), 'e': (-1, 1), 'w': (1, -1), 's': (1, 1)
        }
        
        dx, dy = movements.get(direction, (0, 0))
        x, y = self.player_position

        # Get the number of steps to move based on new position
        new_x, new_y = x + dx, y + dy
        if not (0 <= new_x < HEIGHT and 0 <= new_y < WIDTH):
            self.game_over = True
            return self.score, True
        steps = self.grid[new_x][new_y]

        # Move and calculate score
        move_score = 0
        for i in range(1, steps + 1):
            intermediate_x, intermediate_y = x + i * dx, y + i * dy
            if not (0 <= intermediate_x < HEIGHT and 0 <= intermediate_y < WIDTH) or (intermediate_x, intermediate_y) in self.path:
                self.game_over = True
                return self.score, True
            move_score += self.grid[intermediate_x][intermediate_y]
            self.path.add((intermediate_x, intermediate_y))
            self.grid[intermediate_x][intermediate_y] = 0  # Leave a breadcrumb

        # Update player position
        self.player_position = (x + steps * dx, y + steps * dy)
        self.score += move_score
        return self.score, False

    def reset(self):
        self.grid = np.random.randint(1, 10, size=(HEIGHT, WIDTH))
        self.player_position = (HEIGHT // 2, WIDTH // 2)
        self.grid[self.player_position] = 0
        self.score = 0
        self.path = set([self.player_position])
        self.game_over = False
        return self.grid

    def render(self):
        print(f"Score: {self.score}")
        for i in range(HEIGHT):
            for j in range(WIDTH):
                if (i, j) == self.player_position:
                    print('X', end=' ')
                elif (i, j) in self.path:
                    print('*', end=' ')
                else:
                    print(self.grid[i, j], end=' ')
            print()
        print()

File: dangerstep.cpp
#include <iostream>
#include <vector>
#include <cstdlib>
#include <ctime>

const int WIDTH = 80;
const int HEIGHT = 31;

struct Position {
    int x, y;
};

std::vector<std::vector<int>> initialize_grid() {
    std::vector<std::vector<int>> grid(HEIGHT, std::vector<int>(WIDTH));
    std::srand(static_cast<unsigned int>(std::time(nullptr)));  // Seed for random number generation
    for (auto &row : grid) {
        for (auto &cell : row) {
            cell = std::rand() % 9 + 1;  // Random number from 1 to 9
        }
    }
    return grid;
}

Position get_move_vector(char direction) {
    switch (direction) {
        case 'u': return {-1, 0};
        case 'd': return {1, 0};
        case 'l': return {0, -1};
        case 'r': return {0, 1};
        case 'n': return {-1, -1};
        case 'e': return {-1, 1};
        case 'w': return {1, -1};
        case 's': return {1, 1};
        default: return {0, 0};
    }
}

bool is_valid_position(const Position &pos) {
    return pos.x >= 0 && pos.x < HEIGHT && pos.y >= 0 && pos.y < WIDTH;
}

void print_grid(const std::vector<std::vector<int>> &grid, const std::vector<std::vector<bool>> &path, const Position &player_pos) {
    for (int i = 0; i < HEIGHT; ++i) {
        for (int j = 0; j < WIDTH; ++j) {
            if (i == player_pos.x && j == player_pos.y) {
                std::cout << 'X';
            } else if (path[i][j]) {
                std::cout << '*';
            } else {
                std::cout << grid[i][j];
            }
        }
        std::cout << std::endl;
    }
}

int main() {
    std::vector<std::vector<int>> grid = initialize_grid();
    std::vector<std::vector<bool>> path(HEIGHT, std::vector<bool>(WIDTH, false));
    Position player_pos = {HEIGHT / 2, WIDTH / 2};
    path[player_pos.x][player_pos.y] = true;  // Starting position
    int score = 0;

    print_grid(grid, path, player_pos);
    std::cout << "Score: " << score << "\nMove (u, d, l, r, n, e, s, w): ";

    char command;
    while (std::cin >> command) {
        Position move_vector = get_move_vector(command);
        Position new_pos = {player_pos.x + move_vector.x, player_pos.y + move_vector.y};

        if (!is_valid_position(new_pos)) {
            std::cout << "Invalid move!" << std::endl;
            continue;
        }

        int steps = grid[new_pos.x][new_pos.y];
        int move_score = 0;
        for (int i = 1; i <= steps; ++i) {
            int intermediate_x = player_pos.x + move_vector.x * i;
            int intermediate_y = player_pos.y + move_vector.y * i;
            Position intermediate_pos = {intermediate_x, intermediate_y};

            if (!is_valid_position(intermediate_pos) || path[intermediate_x][intermediate_y]) {
                std::cout << "Game over! Final score: " << score << std::endl;
                return 0;
            }

            move_score += grid[intermediate_x][intermediate_y];
            path[intermediate_x][intermediate_y] = true;
        }

        player_pos = {player_pos.x + move_vector.x * steps, player_pos.y + move_vector.y * steps};
        score += move_score;
        print_grid(grid, path, player_pos);
        std::cout << "Score: " << score << "\nMove (u, d, l, r, n, e, s, w): ";
    }

    return 0;
}